### PR TITLE
chore(core): Stages

### PR DESCRIPTION
**Pull request by eladb**
Stages are self-contained application units that synthesize as a cloud assembly. This change centralizes prepare + synthesis logic into the stage level and changes `App` to extend `Stage`. 

Once `stage.synth()` is called, the stage becomes (practically) immutable. This means that subsequent synths will return the same output.

The cloud assembly produced by stages is nested as an artifact inside another cloud assembly (either the App's top-level assembly) or a child.

Authors: @rix0rrr, @eladb 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/cloud-assembly-schema/scripts/update-schema.sh [hunk 1]
```diff
@@ -1,7 +1,7 @@
 #!/bin/bash
 set -euo pipefail
 scriptsdir=$(cd $(dirname $0) && pwd)
-packagedir=$(realpath ${scriptsdir}/..)
+packagedir=$(cd ${scriptsdir}/.. && pwd)
```
Comment (eladb) [inline packages/@aws-cdk/cloud-assembly-schema/scripts/update-schema.sh:4]: @iliapolo FYI (does not work on a mac)

Comment (iliapolo) [reply packages/@aws-cdk/cloud-assembly-schema/scripts/update-schema.sh:4]: That’s...weird...pretty sure it worked on my Mac 🤷‍♂️

Comment (eladb) [reply packages/@aws-cdk/cloud-assembly-schema/scripts/update-schema.sh:4]: you may have `realpath` installed.

Comment (eladb) [reply packages/@aws-cdk/cloud-assembly-schema/scripts/update-schema.sh:4]: https://github.com/whatwg/html-build/issues/90#:~:text=realpath%20is%20not%20present%20by,Closes%20%2390.&text=realpath%20is%20not%20present%20by%20default%20on%20macOS%2C%20and%20on,not%20support%20%2D%2Drelative%2Dto.

**Pull request merged**